workflow/util/util.go

```diff
@@ -765,6 +765,40 @@ func deletePodNodeDuringRetryWorkflow(wf *wfv1.Workflow, node wfv1.NodeStatus, d
 	return deletedPods, podsToDelete
 }
 
+func containsNode(nodes []string, node string) bool {
+	for _, e := range nodes {
+		if e == node {
+			return true
+		}
+	}
+	return false
+}
+
+func isGroupNode(node wfv1.NodeStatus) bool {
+	return node.Type == wfv1.NodeTypeDAG || node.Type == wfv1.NodeTypeTaskGroup || node.Type == wfv1.NodeTypeStepGroup || node.Type == wfv1.NodeTypeSteps
+}
+
+func resetConnectedParentGroupNodes(oldWF *wfv1.Workflow, newWF *wfv1.Workflow, currentNode wfv1.NodeStatus, resettedParentGroupNodes []string) (*wfv1.Workflow, []string) {
+	currentNodeID := currentNode.ID
+	for {
+		currentNode := oldWF.Status.Nodes[currentNodeID]
+		if !containsNode(resettedParentGroupNodes, currentNodeID) {
+			newWF.Status.Nodes[currentNodeID] = resetNode(*currentNode.DeepCopy())
+			resettedParentGroupNodes = append(resettedParentGroupNodes, currentNodeID)
+			log.Debugf("Resetting connected group node %s", currentNode.Name)
+		}
+		if currentNode.BoundaryID != "" && currentNode.BoundaryID != oldWF.ObjectMeta.Name {
+			parentNode := oldWF.Status.Nodes[currentNode.BoundaryID]
+			if isGroupNode(parentNode) {
+				currentNodeID = parentNode.ID
+			}
```

Comment:
What happens in the implicit else case here? If I'm reading it correctly, it seems currentNodeID doesn't get changed and we end up in an infinite loop, no? Or am I mistaken?